### PR TITLE
GitLab integration

### PR DESCRIPTION
**Pull request by torchiaf**
Signed-off-by: Francesco Torchia <francesco.torchia@suse.com>

- add gitLab APIs
- convert UI components to support generic Git repos
- code refactoring

### Summary
Fixes https://github.com/epinio/ui/issues/173,
Fixes https://github.com/epinio/ui/issues/199

### CODE REVIEW

File: pkg/epinio/detail/applications.vue [hunk 1]
```diff
@@ -178,42 +141,80 @@ export default Vue.extend<Data, any, any, any>({
...
-            commitId:  cur?.sha,
-            author:    cur.author,
-            isChecked: false,
-            date:      cur?.commit.committer.date
-          });
-
-          return acc;
-        }, []);
+    gitType() {
+      const envs = this.value?.envDetails;
```
[the remark this answers is not in the document]
Comment (torchiaf) [reply pkg/epinio/detail/applications.vue]: Updated :+1:

**Comment by torchiaf** [comment]
Thanks for the review @scures , I just added the git Icon based on the git source of the applications, in the `Deployment details` page.
I have opened an issue to request a new icon for GtiLab.

File: pkg/epinio/utils/git.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
+import { EPINIO_APP_ENV_VAR_GIT } from '../types';
+import startCase from 'lodash/startCase';
```
Comment (richard-cox) [inline pkg/epinio/utils/git.ts:2]: Think we're trying t o back out of using lodash. There's a file with some basic string utils `/shell/utils/string.js`. Would `ucFirst` work instead?

File: shell/components/form/GitPicker.vue [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
 import debounce from 'lodash/debounce';
 import { isArray } from '@shell/utils/array';
-import LabeledSelect from '@shell/components/form/LabeledSelect';
-import LabeledInput from '@components/Form/LabeledInput/LabeledInput';
+import LabeledSelect from '@shell/components/form/LabeledSelect.vue';
+import LabeledInput from '@components/Form/LabeledInput/LabeledInput.vue';
 import some from 'lodash/some';
+import { GitUtils } from '@/pkg/epinio/utils/git';
+
+// TODO remove any types
```
Comment (richard-cox) [inline shell/components/form/GitPicker.vue:12]: TODO

File: shell/components/form/GitPicker.vue [hunk 4]
```diff
@@ -140,13 +164,13 @@ export default {
           if (this.oldUsername !== this.selectedAccOrOrg) {
             this.oldUsername = this.selectedAccOrOrg;
 
-            // Resets state, just in case.
+            // Resets state, just in case. TODO to be removed because is called in reset()
```
Comment (richard-cox) [inline shell/components/form/GitPicker.vue:167]: TODO

File: pkg/epinio/utils/git.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        htmlUrl:     data.html_url,
+        name:        data.name
+      }),
+      commit: (data: any) => ({
+        message:  data.commit?.message,
+        htmlUrl:  data.html_url,
+        sha:      getShortHash(data.sha),
+        commitId: data.sha,
+        author:   {
+          name:      data.author.login,
```
Comment (richard-cox) [inline pkg/epinio/utils/git.ts:80]: Bug - create app --> github --> richard-cox --> icons --> master caused a crash. There's a commit that we've fetched that has no auther, this fails.

If this happens on `epinio-dev` (i've got a feeling it does) can you create an issue? If not can you fix in this PR (we should fall back on something sensible)?

File: pkg/epinio/components/application/AppSource.vue [hunk 6]
```diff
@@ -396,8 +397,15 @@ export default Vue.extend<Data, any, any, any>({
...
+    sourceValue() {
+      return {
+        ...this.source.git,
+        type: toLabel(this.type),
+      };
+    },
+
     type() {
-      // There's a bug in the select component which fires off the option ({ value, label}) instead of the value
+      // TODO There's a bug in the select component which fires off the option ({ value, label}) instead of the value
```
Comment (richard-cox) [inline pkg/epinio/components/application/AppSource.vue:408]: TODO

**Review by richard-cox — CHANGES_REQUESTED** [review]
- Sorin called out an issue that might be tracked somewhere else. When the user selects gitlab and then org, branch, etc.... then changes the source type to github... the org/user field remains populated
- This is capture elsewhere - the icon for gitlab is missing

**Comment by torchiaf** [comment]
> - Sorin called out an issue that might be tracked somewhere else. When the user selects gitlab and then org, branch, etc.... then changes the source type to github... the org/user field remains populated
> - This is capture elsewhere - the icon for gitlab is missing

Thanks for the review Richard! Yes, i tracked last week those bugs and I wanted to resolve the  in separated PRs, but since there are other changes to do, I will rather include all in this PR.

File: pkg/epinio/utils/git.ts (hunk 5 above)
Comment (torchiaf) [reply pkg/epinio/utils/git.ts]: It happens only in this PR, just fixed.

File: pkg/epinio/components/application/AppSource.vue (hunk 6 above)
Comment (torchiaf) [reply pkg/epinio/components/application/AppSource.vue:408]: Removed `unSafeType` variable. Do you remember in what case the selected option is of type` ({ value, label})`. I did several tests but I am not able to replicate the bug in this PR.

**Comment by torchiaf** [comment]
@richard-cox I made some code changes to fulfill the requested changes except for the GitLab icon, since we don't have a GitLab icon yet: https://github.com/rancher/dashboard/issues/8407.
Could we open a separate issue for this and postpone the fix to when Manuele will provide the new icon?

File: pkg/epinio/models/application-action.js [hunk 7]
```diff
@@ -165,9 +165,18 @@ export default class ApplicationActionResource extends Resource {
...
         git:  {
-          revision:   source.github.commit,
-          repository: source.github.url,
-          branch:     source.github.branch
+          revision:   source.git.commit,
+          repository: source.git.url,
+          branch:     source.git.branch
+        },
+      };
+    case APPLICATION_SOURCE_TYPE.GIT_LAB:
```
[the remark this answers is not in the document]
Comment (richard-cox) [reply pkg/epinio/models/application-action.js:165]: I think this has introduced a bug. Deploy step fails with `Json: cannot unmarshal string into Go struct field ApplicationOrigin.origin.Kind of type int`. The payload origin's `kind` is `git_lab`, which is an unsupported value (can only be git, see comment for APPLICATION_MANIFEST_SOURCE_TYPE). Looking at the original code it does seem like github was accepted, so not sure what's happening. needs investigation
```
kind: APPLICATION_MANIFEST_SOURCE_TYPE.GIT_HUB,
```

[the remark this answers is not in the document]
Comment (richard-cox) [reply pkg/epinio/models/application-action.js:165]: The same happens when trying to deploy github

[the remark this answers is not in the document]
Comment (richard-cox) [reply pkg/epinio/models/application-action.js:165]: We should set it to GIT and the inner type to source.type

Looking at 
```
    originLabel() {
      return this.value.sourceInfo.kind === APPLICATION_MANIFEST_SOURCE_TYPE.GIT ? ucFirst(this.gitType) : this.value.sourceInfo.label;
    }

    gitType() {
      if (this.envs[APPLICATION_ENV_VAR]) {
        const { type } = JSON.parse(this.envs[APPLICATION_ENV_VAR]) as EPINIO_APP_ENV_VAR_GIT;

        if (type) {
          return type.toLowerCase();
        }
      }

      return 'github';
    },
```

File: pkg/epinio/components/application/AppSource.vue (hunk 6 above)
Comment (richard-cox) [reply pkg/epinio/components/application/AppSource.vue]: Could have been fixed in upstream. If you've tested each source then we can keep your changes

**Review by richard-cox — CHANGES_REQUESTED** [review]
- applications that were previously deployed as `git` have app deployment details as `github` 
- see inline comments for other issues

**Pull request merged**